iroh-net/src/magicsock.rs

```diff
@@ -2518,56 +2420,97 @@ fn bind(
 struct DiscoveredDirectAddrs {
     /// The last set of discovered direct addresses.
-    addrs: Vec<DirectAddr>,
+    addrs: Watchable<BTreeSet<DirectAddr>>,
 
     /// The last time the direct addresses were updated, even if there was no change.
-    updated_at: Option<Instant>,
-}
-
-impl PartialEq for DiscoveredDirectAddrs {
-    fn eq(&self, other: &Self) -> bool {
-        endpoint_sets_equal(&self.addrs, &other.addrs)
-    }
+    ///
+    /// This is only ever None at startup.
+    updated_at: Arc<RwLock<Option<Instant>>>,
 }
 
 impl DiscoveredDirectAddrs {
-    fn new(endpoints: Vec<DirectAddr>) -> Self {
-        Self {
-            addrs: endpoints,
-            updated_at: Some(Instant::now()),
+    /// Updates the direct addresses, returns `true` if they changed, `false` if not.
+    fn update(&self, addrs: BTreeSet<DirectAddr>) -> bool {
+        *self.updated_at.write().unwrap() = Some(Instant::now());
+        let updated = self.addrs.update(addrs).is_ok();
+        if updated {
+            event!(
+                target: "events.net.direct_addrs",
+                Level::DEBUG,
+                addrs = ?self.addrs.get(),
+            );
         }
+        updated
     }
 
-    fn into_iter(self) -> impl Iterator<Item = DirectAddr> {
-        self.addrs.into_iter()
+    fn last_refresh_ago(&self) -> Option<Duration> {
+        self.updated_at
+            .read()
+            .expect("poisoned")
+            .map(|at| at.elapsed())
     }
 
-    fn iter(&self) -> impl Iterator<Item = &DirectAddr> + '_ {
-        self.addrs.iter()
-    }
-
-    fn is_empty(&self) -> bool {
-        self.addrs.is_empty()
+    fn sockaddrs(&self) -> BTreeSet<SocketAddr> {
+        self.addrs.read().iter().map(|da| da.addr).collect()
     }
 
     fn fresh_enough(&self) -> bool {
-        match self.updated_at.as_ref() {
+        match *self.updated_at.read().expect("poisoned") {
             None => false,
             Some(time) => time.elapsed() <= ENDPOINTS_FRESH_ENOUGH_DURATION,
         }
     }
 
     fn to_call_me_maybe_message(&self) -> disco::CallMeMaybe {
-        let my_numbers = self.addrs.iter().map(|ep| ep.addr).collect();
+        let my_numbers = self.addrs.read().iter().map(|da| da.addr).collect();
         disco::CallMeMaybe { my_numbers }
     }
 
-    fn log_direct_addrs_change(&self) {
-        event!(
-            target: "events.net.direct_addrs",
-            Level::DEBUG,
-            addrs = ?self.addrs,
-        );
+    fn updates_stream(&self) -> DirectAddrsStream {
+        DirectAddrsStream {
+            initial: Some(self.addrs.get()),
+            inner: self.addrs.watch().into_stream(),
+        }
+    }
+}
+
+/// Stream returning local endpoints as they change.
+#[derive(Debug)]
+pub struct DirectAddrsStream {
+    initial: Option<BTreeSet<DirectAddr>>,
+    inner: watchable::WatcherStream<BTreeSet<DirectAddr>>,
+}
+
+impl Stream for DirectAddrsStream {
+    type Item = Vec<DirectAddr>;
```

Comment:
`type Item = BTreeSet<DirectAddr>;`?

Reply:
Yeah, this is a fair question.  I left this alone as it changes the public facing API, but it also isn't a big deal to break this I guess.

I really don't know what the right thing is to do here.  Generally direct addresses are going to be few, let's say <10, and in Rust this means a Vec is probably the most efficient data structure for this if you were trying to eek out every bit of perf.

However **logically** it is a set, which is why internally I changed it since it made it easier to reason about and is not on a hot path anyway.  So you're probably right that the same is true for the public API and that conveying that this is a set does make the API more expressive.

Reply:
Personally I think I'm in favor of a set here.

We could even keep the ordering info by having a custom `Ord` impl on `DirectAddr` that first checks the `DirectAddrType`, then the `SocketAddr`. (We might have to only swap the field order for that to happen)

Reply:
A BTreeSet will basically use a vec internally anyway for anything below some threshold. So a BTreeSet with 2 or 3 elements is effectively a vec (single allocation IIRC). It is not something pointer-heavy like a red/black tree.

In any case I think the efficiency is fine in both cases, so using a set to express that duplicates don't make sense here is best in my opinion.

Reply:
Ok, did this now.  Of course this is now a breaking change.  I didn't propagate this change into the `iroh` crate.